cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java

```diff
@@ -0,0 +1,116 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.internal.app.services;
+
+import co.cask.cdap.WordCountApp;
+import co.cask.cdap.app.runtime.ProgramController;
+import co.cask.cdap.app.runtime.ProgramRuntimeService;
+import co.cask.cdap.app.runtime.ProgramRuntimeService.RuntimeInfo;
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.internal.app.services.http.AppFabricTestBase;
+import co.cask.cdap.internal.app.store.DefaultStore;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.cdap.proto.RunRecord;
+
+import org.apache.http.HttpResponse;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+
+/**
+ * Unit test for {@link ProgramLifecycleService}
+ */
+public class ProgramLifecycleServiceTest extends AppFabricTestBase {
+  private static final String WORDCOUNT_APP_NAME = "WordCountApp";
+  private static final String WORDCOUNT_FLOW_NAME = "WordCountFlow";
+
+  private static ProgramLifecycleService programLifecycleService;
+  private static Store store;
+  private static ProgramRuntimeService runtimeService;
+
+  @BeforeClass
+  public static void setup() throws Exception {
+    programLifecycleService = getInjector().getInstance(ProgramLifecycleService.class);
+    store = getInjector().getInstance(DefaultStore.class);
+    runtimeService = getInjector().getInstance(ProgramRuntimeService.class);
+  }
+
+  @Test
+  public void testInvalidFlowRunRecord() throws Exception {
+    // Create App with Flow and the deploy
+    HttpResponse response = deploy(WordCountApp.class, Constants.Gateway.API_VERSION_3_TOKEN, TEST_NAMESPACE1);
+    Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+
+    Id.Program wordcountFlow1 =
+      Id.Program.from(TEST_NAMESPACE1, WORDCOUNT_APP_NAME, ProgramType.FLOW, WORDCOUNT_FLOW_NAME);
+
+    // flow is stopped initially
+    Assert.assertEquals("STOPPED", getProgramStatus(wordcountFlow1));
+
+    // start a flow and check the status
+    startProgram(wordcountFlow1);
+    waitState(wordcountFlow1, ProgramRunStatus.RUNNING.toString());
+
+    // Get the RunRecord
+    List<RunRecord> runRecords = getProgramRuns(wordcountFlow1, ProgramRunStatus.RUNNING.toString());
+    Assert.assertEquals(1, runRecords.size());
+    RunRecord rr = runRecords.get(0);
+
+    // Check the RunRecords status
+    Assert.assertEquals(ProgramRunStatus.RUNNING, rr.getStatus());
+
+    // Lets set the runtime info to off
+    RuntimeInfo runtimeInfo = runtimeService.lookup(wordcountFlow1, RunIds.fromString(rr.getPid()));
+    ProgramController programController = runtimeInfo.getController();
+    programController.stop();
+
+    Thread.sleep(2000);
```

Comment:
Since stop method executing on different thread I found sometimes the status have not been updated properly. Hence adding sleep to make sure it is stopped.